nalgebra-macros/src/lib.rs

```diff
@@ -321,3 +322,401 @@ pub fn point(stream: TokenStream) -> TokenStream {
     };
     proc_macro::TokenStream::from(output)
 }
+
+enum ConcatElem {
+    Zero,
+    Expr(Expr),
+}
+
+impl ConcatElem {
+    fn from_expr(expr: Expr) -> Self {
+        if let Expr::Lit(syn::ExprLit {
+            lit: Lit::Int(ilit),
+            ..
+        }) = &expr
+        {
+            if ilit.base10_digits() == "0" {
+                return ConcatElem::Zero;
+            }
+        }
+        ConcatElem::Expr(expr)
+    }
+
+    fn is_expr(&self) -> bool {
+        matches!(self, ConcatElem::Expr(_))
+    }
+}
+
+/// Constructs a new matrix by concatenating the given matrices.
+///
+/// **Note: Requires the `macros` feature to be enabled (enabled by default)**.
+///
+/// The syntax is similar to the [`matrix!`] and [`dmatrix!`]) macros. However the elements should
+/// be of type `Matrix`, `&Matrix` or be one of the litterals `0` or `1`. The elements of type `Matrix` and `&Matrix` are
+/// concatenated as expected. The litteral `0` is expanded to the zero.Note that at least one element
+/// in each row and column must be an expression of type `Matrix` or `&Matrix`.
+///
+/// All elements in the same row need to have the same number of rows and simillary for the
+/// elements in the same column. This is checked at compile time as long as all elements have
+/// dimensions of type `Const<N>`. If one or more elements have dimensions of type `Dynamic` then
+/// it is checked at runtime that all dimensions match.
+///
+/// If at least one element of a row has `Const<N>` number of rows then that row has a whole will
+/// have `Const<N>` number of rows. However if at least one row has has `Dynamic` number of rows
+/// then the entire matrix will have `Dynamic` number of rows. Same for columns.
```

Comment:
This is a bit confusing; it took me a moment to parse it. I think I understand it after thinking it over, though. We might be able to clarify it a bit by differentiating between "rows" as rows of scalar in a matrix and "rows" as rows of matrices in the stack.

```suggestion
/// If at least one matrix in a row has `Const<...>`  rows, then that row of the stack will have
/// a number of matrix rows known at compile time; if every row of the stack has a number
/// of matrix rows known at compile time then the resulting matrix will have a `Const<...>`
/// number of rows. However, if at least one row of the stack contains only matrices with
/// a dynamic number of rows, then the number of rows in the resulting matrix won't be
/// known at compile time, and so the entire matrix will be given a `Dynamic` number of
/// rows. The same applies to columns.
```